deepspeed/ops/random_ltd/dropping_utils.py

```diff
@@ -0,0 +1,146 @@
+"""
+Copyright 2022 The Microsoft DeepSpeed Team
+"""
+import torch
+
+from ..op_builder import RandomLTDBuilder
+"""
+Returns:
+    sampled_indices: [layers, batch_size, reserved_length]
+    new_mask: [batch_size, 1, reserved_length, reserved_length]
+"""
+
+random_ltd_module = None
+
+
+def gpt_sample_tokens(reserved_length: int,
+                      seq_length: int,
+                      batch_size: int,
+                      layers: int = 1,
+                      device: str = 'cpu',
+                      attn_mask: torch.Tensor = None):
+
+    prob_dist = torch.ones((layers * batch_size, seq_length), device=device)
+    sampled_indices = torch.multinomial(prob_dist, reserved_length)
+
+    sampled_indices = sampled_indices.reshape(layers,
+                                              batch_size,
+                                              reserved_length).to(torch.int32)
+    global random_ltd_module
+    if random_ltd_module is None:
+        random_ltd_module = RandomLTDBuilder().load()
+    sampled_indices = random_ltd_module.token_sort_(sampled_indices, seq_length)
+
+    # Not certain the optimized kernel is actually better here, cause it kind of screws
+    # with alignment right if the sequence length is not divisble by like 16
+    # new_mask = random_ltd_module.mask_gather_gpt(attn_mask, reserved_length)
+    if attn_mask is not None:
+        new_mask = attn_mask[:, :, :reserved_length, :reserved_length]
+    else:
+        new_mask = None
+
+    return sampled_indices, new_mask
+
+
+# TODO(cmikeh2): This probably won't actually work yet, need to sync
```

Comment:
This TODO is out of date and should be removed.

Reply:
Resolved